File: packages/sdk/arc-ts/src/invariants/manifest.ts
import { isValidEd25519PublicKeyHex, publicKeyHexMatches, verifyEd25519Signature } from "./crypto.ts";
import { canonicalizeJson } from "./json.ts";
import { parseJsonText } from "./errors.ts";

export interface MonetaryAmount {
  units: number;
  currency: string;
}

export type PricingModel = "flat" | "per_invocation" | "per_unit" | "hybrid";

export interface ToolPricing {
  pricing_model: PricingModel;
  base_price?: MonetaryAmount;
  unit_price?: MonetaryAmount;
  billing_unit?: string;
}

export interface ToolManifest {
  schema: string;
  server_id: string;
  name: string;
  description?: string;
  version: string;
  tools: Array<{
    name: string;
    description: string;
    input_schema: unknown;
    output_schema?: unknown;
    pricing?: ToolPricing;
    has_side_effects: boolean;
    latency_hint?: "instant" | "fast" | "moderate" | "slow";
  }>;
  required_permissions?: {
    read_paths?: string[];
    write_paths?: string[];
    network_hosts?: string[];
    environment_variables?: string[];
  };
  public_key: string;
}

export interface SignedManifest {
  manifest: ToolManifest;
  signature: string;
  signer_key: string;
}

export interface ManifestVerification {
  structure_valid: boolean;
  signature_valid: boolean;
  embedded_public_key_valid: boolean;
  embedded_public_key_matches_signer: boolean;
}

function validateManifestStructure(manifest: ToolManifest): boolean {
  if (manifest.schema !== "arc.manifest.v1") {
    return false;
  }
  if (manifest.tools.length === 0) {
    return false;
  }

  const seen = new Set<string>();
  for (const tool of manifest.tools) {
    if (seen.has(tool.name)) {
      return false;
    }
    seen.add(tool.name);
  }

  return true;
}

export function parseSignedManifestJson(input: string): SignedManifest {
  return parseJsonText(input);
}

export function signedManifestBodyCanonicalJson(signedManifest: SignedManifest): string {
  return canonicalizeJson(signedManifest.manifest);
}

export function verifySignedManifest(signedManifest: SignedManifest): ManifestVerification {
  const embedded_public_key_valid = isValidEd25519PublicKeyHex(signedManifest.manifest.public_key);

  return {
    structure_valid: validateManifestStructure(signedManifest.manifest),
    signature_valid: verifyEd25519Signature(
      signedManifestBodyCanonicalJson(signedManifest),
      signedManifest.signer_key,
      signedManifest.signature,
    ),
    embedded_public_key_valid,
    embedded_public_key_matches_signer:
      embedded_public_key_valid &&
      publicKeyHexMatches(signedManifest.manifest.public_key, signedManifest.signer_key),
  };
}

export function verifySignedManifestJson(input: string): ManifestVerification {
  return verifySignedManifest(parseSignedManifestJson(input));
}

File: crates/arc-bindings-core/src/manifest.rs
use arc_core::PublicKey;
use arc_manifest::SignedManifest;
use serde::{Deserialize, Serialize};

use crate::Result;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ManifestVerification {
    pub structure_valid: bool,
    pub signature_valid: bool,
    pub embedded_public_key_valid: bool,
    pub embedded_public_key_matches_signer: bool,
}

pub fn parse_signed_manifest_json(input: &str) -> Result<SignedManifest> {
    Ok(serde_json::from_str(input)?)
}

pub fn signed_manifest_body_canonical_json(signed_manifest: &SignedManifest) -> Result<String> {
    arc_core::canonical_json_string(&signed_manifest.manifest).map_err(Into::into)
}

pub fn verify_signed_manifest(signed_manifest: &SignedManifest) -> Result<ManifestVerification> {
    let structure_valid = arc_manifest::validate_manifest(&signed_manifest.manifest).is_ok();
    let embedded_public_key = PublicKey::from_hex(&signed_manifest.manifest.public_key).ok();

    Ok(ManifestVerification {
        structure_valid,
        signature_valid: signed_manifest
            .signer_key
            .verify_canonical(&signed_manifest.manifest, &signed_manifest.signature)?,
        embedded_public_key_valid: embedded_public_key.is_some(),
        embedded_public_key_matches_signer: embedded_public_key
            .as_ref()
            .is_some_and(|key| key == &signed_manifest.signer_key),
    })
}

pub fn verify_signed_manifest_json(input: &str) -> Result<ManifestVerification> {
    let signed_manifest = parse_signed_manifest_json(input)?;
    verify_signed_manifest(&signed_manifest)
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used)]
mod tests {
    use super::verify_signed_manifest;
    use arc_core::Keypair;
    use arc_manifest::{sign_manifest, LatencyHint, ToolDefinition, ToolManifest};

    fn sample_manifest(public_key: String) -> ToolManifest {
        ToolManifest {
            schema: "arc.manifest.v1".to_string(),
            server_id: "srv-bindings-demo".to_string(),
            name: "Bindings Demo".to_string(),
            description: Some("Bindings manifest verification sample".to_string()),
            version: "1.0.0".to_string(),
            tools: vec![ToolDefinition {
                name: "file_read".to_string(),
                description: "Reads a file from the workspace".to_string(),
                input_schema: serde_json::json!({
                    "type": "object",
                    "properties": {
                        "path": { "type": "string" }
                    },
                    "required": ["path"]
                }),
                output_schema: Some(serde_json::json!({
                    "type": "object",
                    "properties": {
                        "content": { "type": "string" }
                    }
                })),
                pricing: None,
                has_side_effects: false,
                latency_hint: Some(LatencyHint::Fast),
            }],
            required_permissions: None,
            public_key,
        }
    }

    #[test]
    fn verify_valid_signed_manifest() {
        let server = Keypair::from_seed(&[21u8; 32]);
        let signed_manifest =
            sign_manifest(&sample_manifest(server.public_key().to_hex()), &server).unwrap();
        let verification = verify_signed_manifest(&signed_manifest).unwrap();

        assert_eq!(
            verification,
            super::ManifestVerification {
                structure_valid: true,
                signature_valid: true,
                embedded_public_key_valid: true,
                embedded_public_key_matches_signer: true,
            }
        );
    }
}
